sdk-core/src/main/java/care/data4life/sdk/tag/TaggingService.kt

```diff
@@ -63,35 +63,48 @@ class TaggingService(
 
     override fun appendDefaultTags(
             resource: Any,
-            oldTags: HashMap<String, String>?
-    ): HashMap<String, String> {
+            oldTags: Tags?
+    ): Tags {
         return when (resource) {
-            is Fhir3Resource -> appendCommonDefaultTags(resource.resourceType, oldTags).also {
-                if (!it.containsKey(TAG_FHIR_VERSION)) {
-                    it[TAG_FHIR_VERSION] = FhirContract.FhirVersion.FHIR_3.version
-                }
-            }
-            is Fhir4Resource -> appendCommonDefaultTags(resource.resourceType, oldTags).also {
-                if (!it.containsKey(TAG_FHIR_VERSION)) {
-                    it[TAG_FHIR_VERSION] = FhirContract.FhirVersion.FHIR_4.version
-                }
-            }
-            else -> appendCommonDefaultTags(null, oldTags).also {
-                it[TAG_APPDATA_KEY] = TAG_APPDATA_VALUE
-            }
+            is Fhir3Resource -> appendCommonDefaultTags(
+                    resource.resourceType,
+                    oldTags
+            ).also { tags -> tagVersion(tags, FhirContract.FhirVersion.FHIR_3) }
+            is Fhir4Resource -> appendCommonDefaultTags(
+                    resource.resourceType,
+                    oldTags
+            ).also { tags -> tagVersion(tags, FhirContract.FhirVersion.FHIR_4) }
+            else -> appendCommonDefaultTags(
+                    null,
+                    oldTags
+            ).also { tags -> tags[TAG_APPDATA_KEY] = TAG_APPDATA_VALUE }
         }
     }
 
-    override fun getTagFromType(
+    override fun tagVersion(
+            tags: Tags,
+            version: FhirContract.FhirVersion
+    ) {
+        if (version == FhirContract.FhirVersion.UNKNOWN) {
+            tags[TAG_APPDATA_KEY] = TAG_APPDATA_VALUE
+        } else {
+            tags[TAG_FHIR_VERSION] = version.version
+        }
+    }
+
+    override fun getTagsFromType(
             resourceType: Class<Any>?
     ): HashMap<String, String> {
-        return hashMapOf<String, String>().also {
-            if (resourceType == null) {
-                it[TAG_APPDATA_KEY] = TAG_APPDATA_VALUE
-            } else {
-                it[TAG_RESOURCE_TYPE] = fhirElementFactory.getFhirTypeForClass(resourceType)!!
-                it[TAG_FHIR_VERSION] = fhirElementFactory.resolveFhirVersion(resourceType).version
+        val tags = hashMapOf<String, String>()
+
+        val version = if (resourceType == null) {
```

Comment:
I think having a AppData type would be more robust than using null

Reply:
Agree, but I will solve it a bit differently.